cmake/linker/armlink/target.cmake

```diff
@@ -108,5 +108,23 @@ function(toolchain_ld_link_elf)
   )
 endfunction(toolchain_ld_link_elf)
 
+# This function will generate the correct CMAKE_C_LINK_EXECUTABLE / CMAKE_CXX_LINK_EXECUTABLE
+# rule to ensure that standard c and runtime libraries are correctly placed
+# and the end of link invocation and doesn't appear in the middle of the link
```

Comment:
Is this explicit ordering working around some gaps in the dependency graph? Would cmake automatically move those libraries to the correct location if the graph was complete?

Reply:
> Is this explicit ordering working around some gaps in the dependency graph?

yes, what we are ensuring here is that we link the std libs at the exact location where they are required and not being relocated by CMake's internal logic.

This is what you usually do in a CMake toolchain file using `CMAKE_<lang>_LINK_EXECUTABLE` when building for bare-metal.

At the same time the use of properties on the linker target follows the design of compiler and linker flags and allow toolchains a uniform way of describing how standard libraries are selected and in which order (even allowing multiple times without de-duplication).

The toolchain can describe a default set, while still allowing Zephyr and Zephyr module to adjust this set (and link order) if there is a need, for example based on Kconfig settings or static libs.
Examples are:
- newlib Toolchain built-in
- picolibc built as module.

Followup to this work will be some proper documentation regarding Zephyr's toolchain infrastructure so that you should be able to create oot toolchain configurations, and even upstream those.

> Would cmake automatically move those libraries to the correct location if the graph was complete?

The problem is that this graph will never be complete.
As you have experienced, then it would require every static library created to know that it wants a specific link, like
`target_link_libraries(foo PUBLIC gcc)`

but that is not scalable and actually makes it harder to create static libs.
Technically we could wrap this for Zephyr libs, but what about libs created by regular `add_library()` CMake calls.
Also such solution would make it much harder to integrate new toolchains in future.

From design perspective, then such library link information is not something a library should ever care about.
This is also a reason why you'd normally place this in a CMake Toolchain file.
CMake toolchain files are fine for well-known project / a project with well-defined boundries but really not flexible enough for a project like Zephyr which tries to span so many use-cases and where things are flexible and can be configured through Kconfig.

Reply:
I think I managed to construct a complete graph in #65434, but the results were not terribly robust. In any case, I think this change is a good step forward.